### ISSUE TITLE
[java] UnusedLocalVariable/UnusedPrivateField false-positive

### ISSUE THREAD

**Issue by boris-petrov** [pmd/pmd#1190]
PMD version `6.4.0`, Java version 10.

```java
var lock = someVariable.readLock();
try (lock) {
	...
}
```

This gives an `UnusedLocalVariable` warning for the `lock` variable which is not correct. This is new syntax available since Java 9.

**Comment by jsotuyod** [comment]
@boris-petrov interesting scenario... non the less I'm having trouble getting a working sample (or at least this is using a new feature I'm not aware of)...

[`Lock`](https://docs.oracle.com/javase/10/docs/api/java/util/concurrent/locks/Lock.html) doesn't implement `Autoclosable` in neither Java 9 nor 10.

Or is this lock one from a lib such as Guava?

I'm trying to figure out if this is an issue of resources not being considered as usages or a problem with the Java grammar itself.

**Comment by boris-petrov** [comment]
@jsotuyod - sorry, maybe I should have been more specific. This is my implementation of Lock that implements `Autoclosable`:

```java
public interface ILock extends AutoCloseable {
	void lock();

	void unlock();

	@Override
	default void close() {
		unlock();
	}
}
```

Does that help?

**Comment by jsotuyod** [comment]
@boris-petrov awesome! So this is merely a logic issue, nothing to do with support for newer Java versions!

**Comment by boris-petrov** [comment]
@jsotuyod - just to mention that the same problem occurs with the `UnusedPrivateField` check. :)
